release_notes/markdown.py

```diff
@@ -0,0 +1,152 @@
+import dataclasses
+import logging
+import typing
+from collections import defaultdict
+
+import release_notes.model as rnm
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class Title:
+    display: str
+    identifiers: list[str]
+
+    def __hash__(self):
+        return hash(self.display)
+
+    def __eq__(self, other):
+        return (
+            other is not None
+            and isinstance(other, Title)
+            and (other == self or hash(other) == hash(self))
+        )
+
+
+def _simple_title(display: str) -> Title:
+    return Title(display, [display.lower()])
+
+
+categories: list[Title] = [
+    Title('⚠️ Breaking Changes', ['action', 'breaking']),
+    Title('📰 Noteworthy', ['noteworthy']),
+    Title('🏃 Others', ['improvement', 'other']),
+    Title('✨ New Features', ['feature']),
+    Title('🐛 Bug Fixes', ['bugfix', 'fix']),
+    Title('📖 Documentation', ['doc']),
+]
+
+target_groups: list[Title] = [
+    _simple_title('USER'),
+    _simple_title('OPERATOR'),
+    _simple_title('DEVELOPER'),
+    _simple_title('DEPENDENCY'),
+]
+
+# identifier -> Title
+categories_by_identifier: dict[str, Title] = {i: k for k in categories for i in k.identifiers}
+target_groups_by_identifier: dict[str, Title] = {i: k for k in target_groups for i in k.identifiers}
+
+
+@dataclasses.dataclass
+class Header:
+    level: int
+    title: str
+
+    def __str__(self):
+        return f"{'#' * self.level} {self.title}\n"  # there should be a new line after the header
+
+
+@dataclasses.dataclass
+class ListItem:
+    level: int
+    text: str
+
+    def __str__(self):
+        return f"{'  ' * (self.level - 1)}- {self.text}"
+
+
+def list_item_from_lines(lines: list[str], start_level: int = 1) -> list[ListItem]:
+    objs = []
+
+    level = start_level
+    last_index: typing.Optional[int] = None
+    for idx, line in enumerate(lines):
+        if line.startswith(' ') or line.startswith('\t'):
+            first_index = line.index(line.lstrip(' \t')[0])
+            if last_index is None or last_index < first_index:
+                last_index = first_index
+                if not idx:
```

Comment:
codestyle: I think this should rather read `idx > 0`

Reply:
I think there might still be a bug here. To me this reads as a function that takes an indented text-block and returns the un-indented elements together with the level they were indented at. But I'm not sure that it is successfully doing that. Will look some more into it tomorrow 🤔